File: ios/HindPharmaApp/Helper/ApiClient.swift
import Foundation

struct APIClient {
 let token:String?
 private func request(path:String,method:String="GET",body:Data?=nil) async throws -> Data { let base=AppConfig.apiBaseURL.absoluteString.trimmingCharacters(in:CharacterSet(charactersIn:"/")); guard let url=URL(string:"\(base)/\(path)") else{throw APIError.server("Invalid API URL")}; var r=URLRequest(url:url);r.httpMethod=method;r.setValue("application/json",forHTTPHeaderField:"Accept");if let token,!token.isEmpty{r.setValue("Bearer \(token)",forHTTPHeaderField:"Authorization")};if let body{r.httpBody=body;r.setValue("application/json",forHTTPHeaderField:"Content-Type")};let(data,response)=try await URLSession.shared.data(for:r);guard let h=response as? HTTPURLResponse,(200...299).contains(h.statusCode) else{let d=(try? JSONSerialization.jsonObject(with:data) as? [String:Any])?["detail"] as? String;throw APIError.server(d ?? "Request failed")};return data }
 func dailyCalling()async throws->[CallingMedical]{try await decode([CallingMedical].self,path:"calling/today")}
 func medicals()async throws->[Medical]{try await decode([Medical].self,path:"medicals")}
 func products(search:String="")async throws->[Product]{try await decode([Product].self,path:"products?search=\(search.addingPercentEncoding(withAllowedCharacters:.urlQueryAllowed) ?? "")")}
 func users()async throws->[TenantUser]{try await decode([TenantUser].self,path:"users")}
 func superAdminDashboard()async throws->SuperAdminDashboard{try await decode(SuperAdminDashboard.self,path:"super-admin/dashboard")}
 func recordCall(_ id:Int)async throws{_=try await request(path:"calling/\(id)/call",method:"POST",body:Data("{}".utf8))}
 func updateCallStatus(_ id:Int,picked:Bool)async throws{_=try await request(path:"calling/\(id)/status",method:"PATCH",body:try JSONSerialization.data(withJSONObject:["is_pick":picked]))}
 func createOrder(medicalID:Int?,items:[OrderItem])async throws{var d:[String:Any]=["items":items.map{["product_id":$0.productId,"quantity":$0.quantity,"price":$0.price as Any] }];if let medicalID{d["medical_id"]=medicalID};_=try await request(path:"orders",method:"POST",body:JSONSerialization.data(withJSONObject:d))}
 private func decode<T:Decodable>(_ type:T.Type,path:String)async throws->T{try JSONDecoder().decode(T.self,from:try await request(path:path))}
}
enum APIError:LocalizedError{case server(String);var errorDescription:String?{if case .server(let v)=self{return v};return nil}}
struct Medical:Codable,Identifiable{let id:Int;let name:String;let area:String?;let phone:String?}
struct Product:Codable,Identifiable{let id:Int;let product_id:String?;let code:String?;let name:String;let unit:String?;let mrp:Double?;let formula:String?;let company:String?}
struct CallingMedical:Codable,Identifiable{let id:Int;let name:String;let phone:String?;let area:String?;let is_call:Int;let is_pick:Int?;let is_not_pick:Int?;var isCalled:Bool{is_call==1}}
struct OrderItem:Codable{let productId:Int;let quantity:Int;let price:Double?;enum CodingKeys:String,CodingKey{case productId="product_id",quantity,price}}
struct TenantUser:Codable,Identifiable{let id:Int;let username:String;let role:String;let name:String?;let phone:String?;let is_active:Int}
struct AdminSummary:Codable,Identifiable{let id:Int;let username:String;let name:String?;let business_name:String;let is_active:Int;let subscription_plan:String?;let subscription_start:String?;let subscription_expiry:String?}
struct SuperAdminDashboard:Codable{let admins:[AdminSummary]}

File: android/app/src/main/java/com/hindtechgroup/hindpharma/Helper/ApiClient.kt
package com.hindtechgroup.hindpharma.Helper

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

class ApiClient(private val token: String? = null) {
    private suspend fun request(path: String, method: String = "GET", body: String? = null): String = withContext(Dispatchers.IO) {
        val connection = (URL("${AppConfig.apiBaseUrl}$path").openConnection() as HttpURLConnection)
        connection.requestMethod = method
        connection.connectTimeout = 10_000
        connection.readTimeout = 10_000
        connection.setRequestProperty("Accept", "application/json")
        token?.takeIf { it.isNotBlank() }?.let { connection.setRequestProperty("Authorization", "Bearer $it") }
        if (body != null) {
            connection.setRequestProperty("Content-Type", "application/json")
            connection.doOutput = true
            connection.outputStream.use { it.write(body.toByteArray()) }
        }
        val code = connection.responseCode
        val stream = if (code in 200..299) connection.inputStream else connection.errorStream
        val response = stream?.bufferedReader()?.use { it.readText() }.orEmpty()
        connection.disconnect()
        if (code !in 200..299) {
            val message = runCatching { JSONObject(response).optString("detail") }.getOrNull()
            throw IllegalStateException(message?.ifBlank { null } ?: "Request failed ($code)")
        }
        response
    }

    suspend fun getMedicals(): List<Medical> = JSONArray(request("/medicals")).let { a -> List(a.length()) { Medical.fromJson(a.getJSONObject(it)) } }
    suspend fun getProducts(search: String = ""): List<Product> = JSONArray(request("/products?search=${java.net.URLEncoder.encode(search, "UTF-8")}")).let { a -> List(a.length()) { Product.fromJson(a.getJSONObject(it)) } }
    suspend fun getDailyCalling(): List<CallingMedical> = JSONArray(request("/calling/today")).let { a -> List(a.length()) { CallingMedical.fromJson(a.getJSONObject(it)) } }
    suspend fun getUsers(): List<TenantUser> = JSONArray(request("/users")).let { a -> List(a.length()) { TenantUser.fromJson(a.getJSONObject(it)) } }
    suspend fun getSuperAdminDashboard(): SuperAdminDashboard {
        val j = JSONObject(request("/super-admin/dashboard")); val a = j.getJSONArray("admins")
        return SuperAdminDashboard(List(a.length()) { AdminSummary.fromJson(a.getJSONObject(it)) })
    }
    suspend fun recordCall(medicalId: Int) = request("/calling/$medicalId/call", "POST", "{}")
    suspend fun updateCallStatus(medicalId: Int, picked: Boolean) = request("/calling/$medicalId/status", "PATCH", JSONObject().put("is_pick", picked).toString())
    suspend fun createOrder(medicalId: Int?, items: List<OrderItem>) = request("/orders", "POST", JSONObject().apply { if (medicalId != null) put("medical_id", medicalId); put("items", JSONArray(items.map { JSONObject().put("product_id", it.productId).put("quantity", it.quantity).put("price", it.price) })) }.toString())
}

data class Medical(val id: Int, val name: String, val area: String?, val phone: String?) { companion object { fun fromJson(j: JSONObject) = Medical(j.optInt("id"), j.optString("name"), j.optString("area").ifBlank { null }, j.optString("phone").ifBlank { null }) } }
data class Product(val id: Int, val productId: String?, val code: String?, val name: String, val unit: String?, val mrp: Double?, val formula: String?, val company: String?) { companion object { fun fromJson(j: JSONObject) = Product(j.optInt("id"), j.optString("product_id").ifBlank { null }, j.optString("code").ifBlank { null }, j.optString("name"), j.optString("unit").ifBlank { null }, if (j.isNull("mrp")) null else j.optDouble("mrp"), j.optString("formula").ifBlank { null }, j.optString("company").ifBlank { null }) } }
data class CallingMedical(val id: Int, val name: String, val phone: String?, val area: String?, val isCall: Boolean, val isPick: Boolean?, val isNotPick: Boolean?) { companion object { fun fromJson(j: JSONObject) = CallingMedical(j.optInt("id"), j.optString("name"), j.optString("phone").ifBlank { null }, j.optString("area").ifBlank { null }, j.optInt("is_call") == 1, if (j.isNull("is_pick")) null else j.optInt("is_pick") == 1, if (j.isNull("is_not_pick")) null else j.optInt("is_not_pick") == 1) } }
data class OrderItem(val productId: Int, val quantity: Int, val price: Double?)
data class TenantUser(val id: Int, val username: String, val role: String, val name: String?, val phone: String?, val isActive: Boolean) { companion object { fun fromJson(j: JSONObject) = TenantUser(j.optInt("id"), j.optString("username"), j.optString("role"), j.optString("name").ifBlank { null }, j.optString("phone").ifBlank { null }, j.optInt("is_active") == 1) } }
data class AdminSummary(val id: Int, val username: String, val name: String?, val businessName: String, val isActive: Boolean, val subscriptionPlan: String?, val subscriptionStart: String?, val subscriptionExpiry: String?) { companion object { fun fromJson(j: JSONObject) = AdminSummary(j.optInt("id"), j.optString("username"), j.optString("name").ifBlank { null }, j.optString("business_name"), j.optInt("is_active") == 1, j.optString("subscription_plan").ifBlank { null }, j.optString("subscription_start").ifBlank { null }, j.optString("subscription_expiry").ifBlank { null }) } }
data class SuperAdminDashboard(val admins: List<AdminSummary>)
